spacer/extract_features.py

```diff
@@ -111,12 +114,50 @@ def feature_dim(self):
 
 class EfficientNetExtractor(FeatureExtractor):
 
+    def __init__(self):
+
+        # Cache models locally.
+        self.modelweighs_path, self.model_was_cashed = download_model(
+            'efficientnetb0_5eps_best.pt')
+
     def __call__(self, im, rowcols):
-        raise NotImplementedError
+
+        start_time = time.time()
+
+        # Set torch parameters
+        torch_params = {'model_type': 'efficientnet',
+                        'model_name': 'efficientnet-b0',
+                        'weights_path': self.modelweighs_path,
+                        'num_class': 1279,
+                        'crop_size': 224,
+                        'batch_size': 10}
+
+        # The imdict data structure needs a label, set to 1, it's not used.
+        rowcollabels = [(row, col, 1) for row, col in rowcols]
+
+        # Crop image to patch list
+        im = np.asarray(im)
```

Comment:
I'd wrap these in a `img_to_patchlist` method. Also, try running your extractor on 
```
image_loc=DataLocation(storage_type='s3',
                                   key='sfq2mr5qbs.jpeg',
                                   bucket_name='spacer-test')
```
and 
```
image_loc=DataLocation(storage_type='s3',
                                   key='kh6dydiix0.jpeg',
                                   bucket_name='spacer-test')
```
These were corner-cases we identified earlier, and for some reason I had to use that do the `im = np.array(pil_im)` conversion twice.

Reply:
I warped the following to the shared `extract_features_utils.py`, `crop_patch` in `extract_features_utils.py` now works as follow so it can be used by both `VGG16CaffeExtractor` and `EfficientNetExtractor`:
- if gray image, convert to RGB and take the first 3 channels only
- crop and return patch list

In terms of the numpy returning empty array for some images, I found a relevant issue: https://github.com/numpy/numpy/issues/11629 and seems like it's PIL issue
```
_ = np.array(im)  # For some images np.array returns an empty array.
im = np.array(im)  # Running it twice fixes this. Don't ask me why.

if len(im.shape) == 2 or im.shape[2] == 1:
    im = gray2rgb(im)
im = im[:, :, :3]
```